README.md
# Serial Port Helper

You can connect any serial port used to read / write data.

## Features

1. Serial Port View;  
   ![](media/View.png)
2. Serial Port Config;  
   ![](media/Config.png)
3. TX / RX;  
   ![](media/TX.png)
   ![](media/RX.png)
4. Send Hex Buffer: Press send button, and then input `@hex` or `@hex:<your buffer byte>` to send Hex Buffer;  
5. Send File: Press send button, and then input `@file` or `@file:<your file path>` to send file.  

## Notice
If you use hex view on RX. It will be output in one line. You can paste this config text in setting.json to make it word wrap.
```json
"[Log]": {
    "editor.wordWrap": "on"
}
```

## Release Notes
### 0.0.15
- Support Last Version VSCode in Windows

### 0.0.14
- Added Init Fail Message.

### 0.0.14
- Added Init Fail Message.
- Support Last Version VSCode in Windows

### 0.0.13
- Added to support Mac M1.

### 0.0.12
- Added support to set keybinding to send data.

### 0.0.11
- Added notice to comment extenion on market.

### 0.0.10
- Fixed send button disappears.

### 0.0.9
- Support Hex View on RX.

### 0.0.8
- Update to support VSCode `1.59.0` on Mac OS(Linux not support yet).

### 0.0.7
- Update to support VSCode `1.59.0` on Windows OS(Mac and Linux not support yet).

### 0.0.6
- Added option `serialPort.enableEscapeCharacte` and support to send escape characte.

### 0.0.5
- Support DIY Baud Rate.

### 0.0.4
- Update preview image size in readme.

### 0.0.3
- Add send hex buffer command.
- Add send file command

### 0.0.2
- Add support Chinese.
- Add options list in config serial port.
- Add remember last serial port config.

### 0.0.1

- First Version, only baise feature.

-----------------------------------------------------------------------------------------------------------
### For more information

* [Serial Port Icon](https://iconscout.com/icons/serial-port) by [Microsoft](https://iconscout.com/contributors/fluent)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "serialport-helper" extension will be documented in this file.

## [0.0.15] - 2022-04-07
- Support Last Version VSCode in Windows

## [0.0.14] - 2022-04-07
- Added Init Fail Message.

## [0.0.13] - 2021-11-30
- Added to support Mac M1.

## [0.0.12] - 2021-08-24
- Added support to set keybinding to send data.
- 
## [0.0.11] - 2021-08-25
- Added notice to comment extenion on market.

## [0.0.10] - 2021-08-24
- Fixed send button disappears.

## [0.0.9] - 2021-08-16
- Support Hex View on RX.

## [0.0.8] - 2021-08-12
- Update to support VSCode `1.59.0` on Mac OS(Linux not support yet).

## [0.0.7] - 2021-08-09
- Update to support VSCode `1.59.0` on Windows OS(Mac and Linux not support yet).

## [0.0.6] - 2021-07-07
- Added option `serialPort.enableEscapeCharacte` and support to send escape characte.

## [0.0.5] - 2021-07-01
- Support DIY Baud Rate.

## [0.0.4] - 2021-07-01
- Update preview image size in readme.

## [0.0.3] - 2021-07-01
- Add send hex buffer command.
- Add send file command

## [0.0.2] - 2021-06-27
- Add support Chinese.
- Add options list in config serial port.
- Add remember last serial port config.

## [0.0.1] - 2021-06-27
- First Version, only baise feature.

src/extension.js
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode = require('vscode');
const { SerialPortProvider } = require('./SerialPort');
const command = require('./comment');

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed

/**
 * @param {vscode.ExtensionContext} context
 */
function activate(context) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.debug('Congratulations, your extension "serialport-helper" is now active!');

	command.init(context);

	context.subscriptions.push(vscode.commands.registerCommand('serialport.connectOrDisconect', command.connectOrDisconect));
	context.subscriptions.push(vscode.commands.registerCommand('serialport.sendEntry', command.sendEntry));
	context.subscriptions.push(vscode.commands.registerCommand('serialport.updateEntry', command.updateEntry));

	const SerialPortsProvider = new SerialPortProvider(context);
	vscode.window.registerTreeDataProvider('serialport', SerialPortsProvider);
	vscode.commands.registerCommand('serialport.refreshEntry', () => SerialPortsProvider.refresh());


}

// this method is called when your extension is deactivated
function deactivate() {}

module.exports = {
	activate,
	deactivate
}


lib/serialport.js
const locale = require('../i18n')();
const vscode = require('vscode');
let SerialPort
try {
    SerialPort = require('node-usb-native').SerialPort;
} catch (error) {
    vscode.window.showWarningMessage(locale['init_fail'])
}

class Module 
{
    constructor(path, options) {
        if (!SerialPort) return;
        this.connected = false;
        options = options || {}
        options.autoOpen = false;
        this.port = new SerialPort(path, options, (err) => {
            this.lasterror = err;
        });
        this.options = options;
        this.path = path;
    }

    onListen(callback) {
        this.port.on('error', (err) => {
            callback(err)
        })

        this.port.on('data', (data) => {
            callback(null, data)
        })
    }

    onChange(callback) {
        this.port.on('open', (data) => {
            callback(true, data)
        })
        this.port.on('close', (data) => {
            callback(false, data)
        })

    }

    open() {
        return new Promise((resolve) => {
            this.port.open((err) => {
                this.lasterror = err;
                resolve(!err);
            })
        })
    }

    close() {
        return new Promise((resolve) => {
            this.port.close((err) => {
                this.lasterror = err;
                resolve(!err);
            })
        })
    }

    send(data) {
        return new Promise((resolve) => {
            this.port.write(data, (err) => {
                this.lasterror = err;
                resolve(!err);
            })
        })
    }

    get isOpen() {
        return this.port.isOpen;
    }

    static search() {
        if (!SerialPort) {
            vscode.window.showWarningMessage(locale['init_fail']);
            return;
        }
        return SerialPort.list();
    }
}

module.exports = Module

src/SerialPort.js
const vscode = require('vscode');
const SerialPort = require('../lib/serialport');
const path = require('path');
const locale = require('../i18n')();


class SerialPortProvider {
	constructor(context) {
		this._onDidChangeTreeData = new vscode.EventEmitter();
		this.refreshPort();
		this.ports = [];
		this.context = context;
	}

	get onDidChangeTreeData() {
		return this._onDidChangeTreeData.event
	}

	async refreshPort() {
		let ports = await SerialPort.search();
		let paths = this.ports.map(p => p.path);
		let portNew = ports.filter(p => paths.indexOf(p.path) < 0);
		paths = ports.map(p => p.path);
		this.ports = this.ports.filter(p => paths.indexOf(p.path) >= 0);
		this.ports = this.ports.concat(portNew.map(p => new SerialPortItem(p, this.context,
			vscode.TreeItemCollapsibleState.Collapsed, () => {
			this._onDidChangeTreeData.fire();
		})))
	}

	refresh() {
		this.refreshPort();
		this._onDidChangeTreeData.fire();
	}

	getTreeItem(element) {
		return element;
	}

	async getChildren(element) {
		if (!element) {
			return await this.getSerialPortList();
		} else {
			console.dir(element);
			return ['baudRate', 'dataBits', 'stopBits', 'parity', 'view_mode'].map(a => new SerialPortAttr(element, a, 
				vscode.TreeItemCollapsibleState.None, () => {
				this._onDidChangeTreeData.fire();
			}))
		}

	}

	async getSerialPortList() {
		return this.ports;
	}
}
class SerialPortItem extends vscode.TreeItem {
	constructor(
		port,
		context,
		collapsibleState,
		changeEvent
	) {
		super(port.path, collapsibleState);
        this.context = context;
        this.contextValue = 'serialport-item';
        this.command= {
			command: 'serialport.connectOrDisconect',
			title: '',
			arguments: [this]
		};
		this.info = port;
		this.changeEvent = changeEvent;
		this.initPort(port.path, this.context.globalState.get(`serialport.${port.path}`) || {
			baudRate: 9600,
			dataBits: 8,
			stopBits: 1,
			parity: 'none',
			view_mode: true
		});
	}

	initPort(path, options) {
		this.options = options;
		let isOpen = this.port && this.port.isOpen;
		if (isOpen) this.port.close();
		this.port = new SerialPort(path, this.options);
		this.port.onChange((isOpen) => {
			this.changeEvent();
			if (isOpen) this.showChannel();
		});
		this.port.onListen((err, data) => {
			if(!err) this.outputData(data);
			else vscode.window.showErrorMessage(`Serial Port [${path}] get some error: ${err.message}`)
		});
		if (isOpen) this.port.open();

		this.context.globalState.update(`serialport.${path}`, this.options);
	}

	outputData(data) {
		if (!this.port.isOpen) return;
		if(!this.options.view_mode) this.outputChannel.append(data.toString());
		else this.outputChannel.append(Array.from(Buffer.from(data, 'utf8')).map(d => ('0' + d.toString(16)).slice(-2)).join(' ') + ' ');
	}

	open() {
		this.outputChannel = vscode.window.createOutputChannel(`${locale['serialport']} [${this.path}]`);
		return this.port.open();
	}

	close() {
		this.outputChannel.clear();
		this.outputChannel.dispose();
		return this.port.close();
	}

	async setting(options) {
		this.initPort(this.port.path, Object.assign(this.options, options));
		this.changeEvent();
		return true;
	}

	get isOpen() {
		return this.port.isOpen;
	}

	get path() {
		return this.info.path;
	}

	showChannel() {
		this.outputChannel.show(true);
	}

	get tooltip() {
		if (!this.info.vendorId) return null;
		return `VID:  ${this.info.vendorId}
PID:  ${this.info.productId}
${locale['manufacturer']}: ${this.info.manufacturer}
${locale['serialNumber']}: ${this.info.serialNumber}
* ${locale['click_to']} ${this.port.isOpen ? locale['disconnect'] : locale['connect']}`;
	}

	get description() {
		return `[${this.port.isOpen ? locale['connect'] : locale['disconnect']}]`;
	}

	get iconPath() {
        return path.join(__filename, '..', '..', 'resources', this.port.isOpen ? 'link.svg' : 'unlink.svg');
    }

	get lasterror() {
		return this.port.lasterror && this.port.lasterror.message;
	}
}

class SerialPortAttr extends vscode.TreeItem {
	constructor(
		port,
		attr,
		collapsibleState,
		changeEvent
	) {
		super(locale[attr], collapsibleState);
        this.contextValue = 'serialport-attr';
        this.command= {
			command: 'serialport.updateEntry',
			title: '',
			arguments: [port, attr]
		};
		this.attr = attr;
		this.port = port;
		this.changeEvent = changeEvent;
	}

	get tooltip() {
		return locale['click_update'];
	}

	get description() {
		if (this.attr == 'view_mode') return !this.port.options[this.attr] ? locale['string_mode'] : locale['hex_mode'];
		return this.port.options[this.attr].toString();
	}

	get iconPath() {
        return {
			light: path.join(__filename, '..', '..', 'resources', 'light', 'setting.svg'),
			dark: path.join(__filename, '..', '..', 'resources', 'dark', 'setting.svg')
		};
    }
}

module.exports = { SerialPortItem, SerialPortProvider};

src/serialportItem.js
const vscode = require('vscode');
const SerialPort = require('../lib/serialport');
const path = require('path');

class SerialPortItem extends vscode.TreeItem {

	constructor(
		label,
		collapsibleState,
		command
	) {
		super(label, collapsibleState);
        this.port = new SerialPort(label);
        this.contextValue = 'serialport-item';
        this.command= command;
	}

	get tooltip() {
		return `${this.label}(${this.port.open ? 'connected' : 'disconected'})`;
	}

	get description() {
		return this.version;
	}

	get iconPath() {
        return {
            light: path.join(__filename, '..', '..', 'resources', 'light', this.port.open ? 'link.svg' : 'unlink.svg'),
            dark: path.join(__filename, '..', '..', 'resources', 'dark', this.port.open ? 'link.svg' : 'unlink.svg')
        };
    }
}

module.exports = SerialPortItem;

src/comment.js
const vscode = require('vscode');
const locale = require('../i18n')();
const fs = require('fs');
const open = require('child_process');

let attrOptions = {
	baudRate: [115200, 57600, 38400, 19200, 9600, 4800, 2400, 1800, 1200, 600, 300, 200, 150, 134, 110, 75, 50, 'DIY'],
	dataBits: [8, 7, 6, 5],
	stopBits: [1, 2],
	parity: ['none', 'even', 'mark', 'odd', 'space']
};

let usetimes = 0;

let ports = {};

async function getSendData(data, path) {
	if (!data) return data;
	// send file
	if (data.startsWith('@file')) {
		let filepath = data.replace(/^@file:/, '');
		if(!fs.existsSync(filepath)) {
			filepath = await vscode.window.showOpenDialog({ canSelectMany: false });
			if (filepath.length == 0) return null;
			filepath = filepath[0].fsPath;
		}

		return fs.readFileSync(filepath);
	}

	// send hex
	if (data.startsWith('@hex')) {
		data = data.replace(/^@hex:/, '');
		if (data.match(/^\s*([0-9A-Fa-f]{2}\s+)*[0-9A-Fa-f]{2}\s*$/g) == null) {
			data = await vscode.window.showInputBox({ prompt: locale['send_hex_title'].replace(/{{path}}/, path) });
			if (!data) return data;
			if (data.match(/^\s*([0-9A-Fa-f]{2}\s+)*[0-9A-Fa-f]{2}\s*$/g) == null) {
				vscode.window.showErrorMessage(locale['not_invalid_hex']);
				return null;
			}
		}
		data = data.match(/[0-9A-Fa-f]{2}\s*/g).map(d => parseInt(d.trim(), 16));
		return Buffer.from(data);
	}

	if (vscode.workspace.getConfiguration().get('serialPort.enableEscapeCharacte')) {
		try { data = JSON.parse(`\"${data.replace(/"/g, `\\"`)}\"`); } catch(ex) {}
	}

	return data;
}

function getExecCommand() {
    let cmd = 'start';
    if (process.platform == 'win32') {
        cmd = 'start';
    } else if (process.platform == 'linux') {
        cmd = 'xdg-open';
    } else if (process.platform == 'darwin') {
        cmd = 'open';
    }
    
    return `${cmd} https://marketplace.visualstudio.com/items?itemName=hancel.serialport-helper`
}

function confirm(message, options) {
    return new Promise((resolve, reject) => {
        return vscode.window.showInformationMessage(message, ...options).then(resolve);
    });
}

function noticeComment() {
    let notice = context.globalState.get('notice');
    if (!notice && usetimes > 80) {
        confirm(locale['like.extension'], [locale['like.ok'], locale['like.no'], locale['like.later']])
            .then((option) => {
                switch(option) {
                    case locale['like.ok']:
                        open.exec(getExecCommand());
                        context.globalState.update('notice', true);
                        break;
                    case locale['like.no']:
                        context.globalState.update('notice', true);
                        break;
                    case locale['like.later']:
                        usetimes = 50;
                        context.globalState.update('usetimes', usetimes);
                        context.globalState.update('notice', false);
                        break;
                }
            })
            .catch(e => console.log(e));
    } else if(!notice) {
        context.globalState.update('usetimes', ++usetimes);
    }
}

module.exports = {
	init(cxt) {
		usetimes = cxt.globalState.get('usetimes') || 0;
	},

    async connectOrDisconect(port) {
		let ret = false, isOpen = port.isOpen;
		if (isOpen) ret = await port.close();
		else{
			ret = await port.open();
			ports[port.path] = port;
			noticeComment()
		}

		if (!ret) vscode.window.showErrorMessage(port.lasterror);
    },

	async updateEntry(port, attr) {
		if (attr == 'view_mode') {
			port.options[attr] = !port.options[attr]
			await port.setting({ [attr]: port.options[attr] });
			return;
		}
		let data = port.options[attr];
		data = await vscode.window.showQuickPick(attrOptions[attr].map(a => ({ label: a.toString() })), 
			{ title: locale['update_title'].replace(/{{path}}/, port.path).replace(/{{attr}}/, locale[attr]) });
		if (!data) return;
		if (data.label == 'DIY') data.label = await vscode.window.showInputBox({ 
			prompt: locale['update_title'].replace(/{{path}}/, port.path).replace(/{{attr}}/, locale[attr]),
			validateInput(value) {
				return !isNaN(value) ? null : `${value} ${locale['is_not_number']}`;
			}
		});
		data = attr != 'parity' ? parseInt(data.label) : data.label;
		let ret = true;
		await port.setting({ [attr]: data });
		if (ret) {
			port.options[attr] = data;
		} else vscode.window.showErrorMessage(`Update ${attr} of ${port.path} was failed. Error Message: ${port.lasterror}`);
	},

	async sendEntry(port) {
		if (typeof port == 'string') {
			if (ports[port] === undefined) {
				vscode.window.showWarningMessage(`${port} was not found or not connected yet.`);
				return;
			}
			port = ports[port];
		}
		let data = await vscode.window.showInputBox({ prompt: locale['send_title'].replace(/{{path}}/, port.path) });
		data = await getSendData(data, port.path);
		if (!data) return;
		let ret = await port.port.send(data);
		if (ret) vscode.window.showInformationMessage(locale['send_success'].replace(/{{path}}/, port.path));
		else vscode.window.showErrorMessage(`Send Data to ${port.path} was failed. Error Message: ${port.lasterror}`);
	}
}